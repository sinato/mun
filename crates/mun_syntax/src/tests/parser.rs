use crate::SourceFile;

fn snapshot_test(text: &str) {
    let text = text.trim().replace("\n    ", "\n");
    let file = SourceFile::parse(&text);
    insta::assert_snapshot!(insta::_macro_support::AutoName, file.debug_dump(), &text);
}

#[test]
fn empty() {
    snapshot_test(r#""#);
}

#[test]
fn function() {
    snapshot_test(
        r#"
    // Source file comment

    // Comment that belongs to the function
    fn a() {}
    fn b(value:number) {}
    pub fn d() {}
    pub fn c()->never {}
    fn b(value:number)->number {}"#,
    );
}

#[test]
fn block() {
    snapshot_test(
        r#"
    fn foo() {
        let a;
        let b:i32;
        let c:string;
    }"#,
    );
}

#[test]
fn literals() {
    snapshot_test(
        r#"
    fn foo() {
        let a = true;
        let b = false;
        let c = 1;
        let d = 1.12;
        let e = "Hello, world!"
    }
    "#,
    );
}

#[test]
fn struct_def() {
    snapshot_test(
        r#"
    struct Foo      // error: expected a ';', or a '{'
    struct Foo;
    struct Foo;;    // error: expected a declaration
    struct Foo {}
    struct Foo {};
    struct Foo {,}; // error: expected a field declaration
    struct Foo {
        a: f64,
    }
    struct Foo {
        a: f64,
        b: i32,
    };
    struct Foo()
    struct Foo();
    struct Foo(,);  // error: expected a type
    struct Foo(f64)
    struct Foo(f64,);
    struct Foo(f64, i32)
    "#,
    )
}

#[test]
fn unary_expr() {
    snapshot_test(
        r#"
    fn foo() {
        let a = --3;
        let b = !!true;
    }
    "#,
    )
}

#[test]
fn binary_expr() {
    snapshot_test(
        r#"
    fn foo() {
        let a = 3+4*5
        let b = 3*4+10/2
    }
    "#,
    )
}

#[test]
fn expression_statement() {
    snapshot_test(
        r#"
    fn foo() {
        let a = "hello"
        let b = "world"
        let c
        b = "Hello, world!"
        !-5+2*(a+b);
        -3
    }
    "#,
    )
}

#[test]
fn function_calls() {
    snapshot_test(
        r#"
    fn bar(i:number) { }
    fn foo(i:number) {
      bar(i+1)
    }
    "#,
    )
}

#[test]
fn patterns() {
    snapshot_test(
        r#"
    fn main(_:number) {
       let a = 0;
       let _ = a;
    }
    "#,
    )
}

#[test]
fn arithmetic_operands() {
    snapshot_test(
        r#"
    fn main() {
        let _ = a + b;
        let _ = a - b;
        let _ = a * b;
        let _ = a / b;
        let _ = a % b;
        let _ = a << b;
        let _ = a >> b;
        let _ = a & b;
        let _ = a | b;
        let _ = a ^ b;
    }
    "#,
    )
}

#[test]
fn assignment_operands() {
    snapshot_test(
        r#"
    fn main() {
        let a = b;
        a += b;
        a -= b;
        a *= b;
        a /= b;
        a %= b;
        a <<= b;
        a >>= b;
        a &= b;
        a |= b;
        a ^= b;
    }
    "#,
    )
}

#[test]
fn compare_operands() {
    snapshot_test(
        r#"
    fn main() {
        let _ = a == b;
        let _ = a == b;
        let _ = a != b;
        let _ = a < b;
        let _ = a > b;
        let _ = a <= b;
        let _ = a >= b;
    }
    "#,
    )
}

#[test]
fn logic_operands() {
    snapshot_test(
        r#"
    fn main() {
        let _ = a || b;
        let _ = a && b;
    }
    "#,
    )
}

#[test]
fn if_expr() {
    snapshot_test(
        r#"
    fn bar() {
        if true {};
        if true {} else {};
        if true {} else if false {} else {};
        if {true} {} else {}
    }
    "#,
    );
}

#[test]
fn type_alias() {
    snapshot_test(
        r#"
    fn foo() {
        type Typename = i32;
        type Typename2 = Typename;
        let a: Typename2 = 1;
        let b: i32;
    }
    "#,
    )
}

#[test]
fn block_expr() {
    snapshot_test(
        r#"
    fn bar() {
        {3}
    }
    "#,
    );
}

#[test]
fn return_expr() {
    snapshot_test(
        r#"
    fn foo() {
        return;
        return 50;
    }
    "#,
    )
}

#[test]
fn loop_expr() {
    snapshot_test(
        r#"
    fn foo() {
        loop {}
    }"#,
    )
}

#[test]
fn break_expr() {
    snapshot_test(
        r#"
    fn foo() {
        break;
        if break { 3; }
        if break 4 { 3; }
    }
    "#,
    )
}

#[test]
fn while_expr() {
    snapshot_test(
        r#"
    fn foo() {
        while true {};
        while { true } {};
    }
    "#,
    )
}

#[test]
fn struct_lit() {
    snapshot_test(
        r#"
    fn foo() {
        U;
        S {};
        S { x, y: 32, };
        S { x: 32, y: 64 };
        TupleStruct { 0: 1 };
        T(1.23);
        T(1.23, 4,)
    }
    "#,
    )
}

#[test]
fn struct_field_index() {
    snapshot_test(
        r#"
    fn main() {
        foo.a
        foo.a.b
        foo.0
        foo.0.1
        foo.10
        foo.01  // index: .0
        foo.0 1 // index: .0
        foo.a.0
    }
    "#,
    )
}

#[test]
fn memory_type_specifier() {
    snapshot_test(
        r#"
    struct Foo {};
    struct(gc) Baz {};
    struct(value) Baz {};
    struct() Err1 {};    // error: expected memory type specifier
    struct(foo) Err2 {}; // error: expected memory type specifier
    "#,
    )
}

#[test]
fn visibility() {
    snapshot_test(
        r#"
    pub struct Foo {};
    pub(package) struct(gc) Baz {};
    pub(super) fn foo() {}
    pub(package) fn bar() {}
    pub fn baz() {}
    "#,
    )
}

#[test]
fn extern_fn() {
    snapshot_test(
        r#"
    pub extern fn foo();
    "#,
    )
}
