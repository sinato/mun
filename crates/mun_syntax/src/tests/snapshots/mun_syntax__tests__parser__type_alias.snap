---
source: crates/mun_syntax/src/tests/parser.rs
expression: "fn foo() {\n    type Typename = i32;\n    type Typename2 = Typename;\n    let a:Typename2;\n}"
---
SOURCE_FILE@[0; 89)
  FUNCTION_DEF@[0; 89)
    FN_KW@[0; 2) "fn"
    WHITESPACE@[2; 3) " "
    NAME@[3; 6)
      IDENT@[3; 6) "foo"
    PARAM_LIST@[6; 8)
      L_PAREN@[6; 7) "("
      R_PAREN@[7; 8) ")"
    WHITESPACE@[8; 9) " "
    BLOCK_EXPR@[9; 89)
      L_CURLY@[9; 10) "{"
      WHITESPACE@[10; 15) "\n    "
      TYPE_ALIAS_DEF@[15; 35)
        TYPE_KW@[15; 19) "type"
        WHITESPACE@[19; 20) " "
        NAME@[20; 28)
          IDENT@[20; 28) "Typename"
        WHITESPACE@[28; 29) " "
        EQ@[29; 30) "="
        WHITESPACE@[30; 31) " "
        PATH_TYPE@[31; 34)
          PATH@[31; 34)
            PATH_SEGMENT@[31; 34)
              NAME_REF@[31; 34)
                IDENT@[31; 34) "i32"
        SEMI@[34; 35) ";"
      WHITESPACE@[35; 40) "\n    "
      TYPE_ALIAS_DEF@[40; 66)
        TYPE_KW@[40; 44) "type"
        WHITESPACE@[44; 45) " "
        NAME@[45; 54)
          IDENT@[45; 54) "Typename2"
        WHITESPACE@[54; 55) " "
        EQ@[55; 56) "="
        WHITESPACE@[56; 57) " "
        PATH_TYPE@[57; 65)
          PATH@[57; 65)
            PATH_SEGMENT@[57; 65)
              NAME_REF@[57; 65)
                IDENT@[57; 65) "Typename"
        SEMI@[65; 66) ";"
      WHITESPACE@[66; 71) "\n    "
      LET_STMT@[71; 87)
        LET_KW@[71; 74) "let"
        WHITESPACE@[74; 75) " "
        BIND_PAT@[75; 76)
          NAME@[75; 76)
            IDENT@[75; 76) "a"
        COLON@[76; 77) ":"
        PATH_TYPE@[77; 86)
          PATH@[77; 86)
            PATH_SEGMENT@[77; 86)
              NAME_REF@[77; 86)
                IDENT@[77; 86) "Typename2"
        SEMI@[86; 87) ";"
      WHITESPACE@[87; 88) "\n"
      R_CURLY@[88; 89) "}"

